customer-service/src/main/java/io/eventuate/examples/tram/ordersandcustomers/customers/service/CustomerService.java

```diff
@@ -66,54 +70,63 @@ public CustomerService(CustomerRepository customerRepository,
       logger.info("reserving credit (orderId: {}, customerId: {})", orderId, customerId);
 
       CustomerCreditReservedEvent customerCreditReservedEvent =
               new CustomerCreditReservedEvent(orderId);
 
-      Mono<CreditReservation> creditReservation =
-              creditReservationRepository.save(new CreditReservation(customerId, orderId, orderTotal.getAmount()));
+      CreateOrderSagaStepSucceededEvent createOrderSagaStepSucceededEvent =
+              new CreateOrderSagaStepSucceededEvent("Customer Service", new OrderDetails(customerId, orderTotal));
 
-      return creditReservation.flatMap(rc ->
-              domainEventPublisher
-                      .publish(Customer.class,
-                              customer.getId(),
-                              Collections.singletonList(customerCreditReservedEvent))
-                      .collectList());
+      return creditReservationRepository
+              .save(new CreditReservation(customerId, orderId, orderTotal.getAmount()))
+              .flatMap(notUsed -> publishCreditReservationEvents(customerId, customerCreditReservedEvent, orderId, createOrderSagaStepSucceededEvent));
     } else {
       logger.info("handling credit reservation failure (orderId: {}, customerId: {})", orderId, customerId);
 
       CustomerCreditReservationFailedEvent customerCreditReservationFailedEvent =
               new CustomerCreditReservationFailedEvent(orderId);
 
-      return domainEventPublisher.publish(Customer.class,
-              customer.getId(),
-              Collections.singletonList(customerCreditReservationFailedEvent))
-              .collectList();
+      CreateOrderSagaStepFailedEvent createOrderSagaStepFailedEvent =
+                        new CreateOrderSagaStepFailedEvent("Customer Service", new OrderDetails(customerId, orderTotal));
+
+      return publishCreditReservationEvents(customerId, customerCreditReservationFailedEvent, orderId, createOrderSagaStepFailedEvent);
     }
   }
 
-  private Mono<List<Message>> handleNotExistingCustomer(long orderId, long customerId) {
+  private Mono<List<Message>> publishCreditReservationEvents(Long customerId, DomainEvent customerEvent, String orderId, DomainEvent orderSagaEvent) {
+
+    return domainEventPublisher.publish(
+            createEvents()
+                    .aggregateType(Customer.class).aggregateId(customerId).event(customerEvent).next()
+                    .aggregateType("io.eventuate.examples.tram.ordersandcustomers.orders.domain.Order").aggregateId(orderId).event(orderSagaEvent)
+                    .build()
+    );
+
+  }
+
+  private Mono<List<Message>> handleNotExistingCustomer(String orderId, long customerId) {
     return Mono.defer(() ->
-      domainEventPublisher
-              .publish(Customer.class,
-                      customerId,
-                      Collections.singletonList(new CustomerValidationFailedEvent(orderId)))
-              .collectList());
+      domainEventPublisher.publish(
```

Comment:
This can be simplified by 

* combining the Builder and Publishing APIs - build() -> publish() since the point of building is to publish.
* eliminating next() - simply call aggregateType() again

```
    return domainEventPublisher
                    .aggregateType(Customer.class).aggregateId(customerId).event(customerEvent)
                    .aggregateType("io.eventuate.examples.tram.ordersandcustomers.orders.domain.Order")
                                .aggregateId(orderId).event(orderSagaEvent)
                    .publish()
    );

```